src/didcomm/encrypted.rs

```diff
@@ -1,283 +1,335 @@
+    type VerifyError = DidCommEncryptedServiceVerifyError<R::FindIdentifierError>;
+    async fn generate(
         &self,
-        from_did: &str,
-        to_did: &str,
+        model: VerifiableCredentials,
         from_keyring: &KeyPairing,
-        message: &Value,
+        to_did: &str,
         metadata: Option<&Value>,
-        issuance_date: DateTime<Utc>,
-    ) -> Result<DIDCommMessage, DIDCommEncryptedServiceGenerateError> {
-        // NOTE: recipient to
-        let did_document =
-            self.vc_service.did_repository.find_identifier(to_did).await?.ok_or_else(|| {
-                DIDCommEncryptedServiceGenerateError::DIDNotFound(to_did.to_string())
-            })?;
-
-        let public_keys = did_document.did_document.public_key.ok_or_else(|| {
-            DIDCommEncryptedServiceGenerateError::DidPublicKeyNotFound(to_did.to_string())
-        })?;
-
-        // FIXME: workaround
-        if public_keys.len() != 1 {
-            return Err(anyhow::anyhow!("public_keys length must be 1").into());
-        }
-
-        let public_key = public_keys[0].clone();
-
-        let other_key = keyring::secp256k1::Secp256k1::from_jwk(&public_key.public_key_jwk)?;
-
-        // NOTE: ecdh
-        let shared_key = runtime::secp256k1::Secp256k1::ecdh(
-            &from_keyring.sign.get_secret_key(),
-            &other_key.get_public_key(),
-        )?;
-
-        let sk = StaticSecret::from(array_ref!(shared_key, 0, 32).to_owned());
-        let pk = PublicKey::from(&sk);
-
-        // NOTE: message
-        let body = self.vc_service.generate(from_did, from_keyring, message, issuance_date)?;
-        let body = serde_json::to_string(&body).context("failed to serialize")?;
-
-        let mut message =
-            Message::new().from(from_did).to(&[to_did]).body(&body).map_err(|e| {
-                anyhow::anyhow!("Failed to initialize message with error = {:?}", e)
-            })?;
-
-        // NOTE: Has attachment
-        if let Some(value) = metadata {
-            let id = cuid::cuid2();
-
-            // let media_type = "application/json";
-            let data = AttachmentDataBuilder::new()
-                .with_link(&self.attachment_link)
-                .with_json(&value.to_string());
-
-            message.append_attachment(
-                AttachmentBuilder::new(true).with_id(&id).with_format("metadata").with_data(data),
-            )
-        }
-
-        let seal_signed_message = message
-            .as_jwe(&CryptoAlgorithm::XC20P, Some(pk.as_bytes().to_vec()))
-            .seal_signed(
-                sk.to_bytes().as_ref(),
-                Some(vec![Some(pk.as_bytes().to_vec())]),
-                SignatureAlgorithm::Es256k,
-                &from_keyring.sign.get_secret_key(),
-            )
-            .map_err(|e| {
-                DIDCommEncryptedServiceGenerateError::EncryptFailed(anyhow::Error::msg(
-                    e.to_string(),
-                ))
-            })?;
```

Comment:
This procedure is totaly insane.

First, secp256k1 is not adapted to DIDComm encryption.

See:
- https://identity.foundation/didcomm-messaging/spec/#curves-and-content-encryption-algorithms

Second, the didcomm-rs api does the ecdh scheme. However, this implementation does the ecdh scheme before calling the didcomm-rs api.
In other words, this implementation performs the ecdh scheme twice.

See:
- https://github.com/nodecross/didcomm-rs/blob/b206c57b85165dec6e4e5a8f734b27ad5ef7befe/src/messages/message.rs#L633
  - https://github.com/nodecross/didcomm-rs/blob/b206c57b85165dec6e4e5a8f734b27ad5ef7befe/src/messages/message.rs#L522
    - https://github.com/nodecross/didcomm-rs/blob/b206c57b85165dec6e4e5a8f734b27ad5ef7befe/src/messages/helpers/encryption.rs#L140
      - https://github.com/nodecross/didcomm-rs/blob/b206c57b85165dec6e4e5a8f734b27ad5ef7befe/src/messages/helpers/encryption.rs#L366

<!--
As a result, the recipient and sender have the same secret/public key.
-->